### ISSUE TITLE
[System Test][5.2.1.0] unable to do  s3 cp with obj lagers than 1T

### ISSUE THREAD

**Issue by javieramirez1** [noobaa/noobaa-core#8151]
### Environment info
- NooBaa Version: **VERSION**
- Platform: **Kubernetes 1.14.1 | minikube 1.1.1 | OpenShift 4.1 | other: specify**
[root@c83f2-dan9 ~]# rpm -qa | grep  noobaa
noobaa-core-5.15.3-20240529.el9.x86_64
[root@c83f2-dan9 ~]# rpm -qa | grep  mms3
gpfs.mms3-5.2.1-0.240604.102411.el9.x86_64
[root@c83f2-dan9 ~]# mmdiag
Current GPFS build: "5.2.1.0 240604.102411".
Built on Jun  4 2024 at 11:44:08
Running 3 days 3 hours 34 minutes 29 secs, pid 1608736

Run mmdiag --help for more options
Run mmdiag --help for more options

[root@c83f2-dan9 ~]# mms3 config list

S3 Configuration:

=======================

ALLOW_HTTP : false
DEBUGLEVEL : default
ENABLEMD5 : false
ENDPOINT_FORKS : 16
ENDPOINT_PORT : 6001
ENDPOINT_SSL_PORT : 6443
GPFSDLPATH : /usr/lpp/mmfs/lib/libgpfs.so
NC_MASTER_KEYS_GET_EXECUTABLE : /usr/lpp/mmfs/bin/cess3_key_get
NC_MASTER_KEYS_PUT_EXECUTABLE : /usr/lpp/mmfs/bin/cess3_key_put
NC_MASTER_KEYS_STORE_TYPE : executable
NSFS_DIR_CACHE_MAX_DIR_SIZE : 536870912
NSFS_DIR_CACHE_MAX_TOTAL_SIZE : 1073741824
NSFS_NC_CONFIG_DIR_BACKEND : GPFS
NSFS_NC_STORAGE_BACKEND : GPFS
UVTHREADPOOLSIZE : 16

=======================
### Actual behavior
1. 

upload failed: - to s3://bucket1/zeros_file_10TB An error occurred (InvalidArgument) when calling the UploadPart operation: Invalid Argument
[root@c83f2-dan9 bucket1]# grep -i invalid /var/log/noobaa.log
2024-05-20T02:23:34.616569-04:00 c83f2-dan9 node[1228864]: [nsfs/1228864] [ERROR] CONSOLE:: ENDPOINT CLIENT ERROR - REPLY WITH BAD REQUEST [Error: Parse Error: Unexpected start char in url] { bytesParsed: 4, code: 'HPE_INVALID_URL', reason: 'Unexpected start char in url', rawPacket: <Buffer 47 45 54 20 2e 2e 2f 2e 2e 2f 2e 2e 2f 2e 2e 2f 2e 2e 2f 2e 2e 2f 2e 2e 2f 2e 2e 2f 2e 2e 2f 2e 2e 2f 2e 2e 2f 2e 2e 2f 77 69 6e 64 6f 77 73 2f 77 69 ... 334 more bytes> }
### Expected behavior
1.1024+0 records in
1024+0 records out
1073741824 bytes (1.1 GB, 1.0 GiB) copied, 11.0929 s, 96.8 MB/s
urllib3/connectionpool.py:1045: InsecureRequestWarning: Unverified HTTPS request is being made to host '172.20.100.60'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
urllib3/connectionpool.py:1045: InsecureRequestWarning: Unverified HTTPS request is being made to host '172.20.100.60'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
                                                                                                                                                                                            real    0m11.590s
user    0m5.647s
sys     0m3.210s

### Steps to reproduce
1.dd if=/dev/zero bs=1M count=$((10*1024*1024)) | AWS_ACCESS_KEY_ID=Nk6lx6msTmKXUM04Chu9 AWS_SECRET_ACCESS_KEY=CzD+xVImh3WocHG1Ag670S+WG0+lqZ3GZMrQ1BMD aws  --no-verify-ssl --endpoint https://172.20.100.60:6443/ s3 cp - s3://bucket1/zeros_file_10TB

**Comment by guymguym** [comment]
@javieramirez1 this seems like a client issue - I decoded the raw packet that the server received and it's not a valid url -
```
> node -p 'Buffer.from("47 45 54 20 2e 2e 2f 2e 2e 2f 2e 2e 2f 2e 2e 2f 2e 2e 2f 2e 2e 2f 2e 2e 2f 2e 2e 2f 2e 2e 2f 2e 2e 2f 2e 2e 2f 2e 2e 2f 77 69 6e 64 6f 77 73 2f 77 69".split(" ").map(i => Number("0x"+i))).toString()'

GET ../../../../../../../../../../../../windows/wi
```

**Comment by javieramirez1** [comment]
@guymguym But what is the difference with small objects, because there are no issues?

**Comment by guymguym** [comment]
@javieramirez1 the difference is the size :) 

my point is that I think it might be a problem of the client and not the server, but lets check that to make sure.

run the following client tool we have - it uploads random generated data without saving it to disk first so there is no source file needed, notice that the sizes are in MB so this example will upload 1TB in 1024 parts of 1024MB each (the total number of parts cannot exceed 10,000).

```
> cd /usr/local/noobaa-core
> bin/node src/tools/s3cat --endpoint http://localhost:6001 --access_key XXX --secret_key YYY --bucket BBB --upload --size $((1024*1024)) --part_size 1024 --concur 1
```

this tool shows only the speed of writing, but not real progress, so you can monitor the progress by looking at the buckets internal dir like so:
```
> ls -lh /gpfs/path/to/bucket/.noobaa-nsfs_*/multipart-uploads/*/parts-size-*
```

**Comment by javieramirez1** [comment]
@guymguym  in my clients i dont have the nooba-core dir Do I copy it from my dan nodes?

**Comment by guymguym** [comment]
no, just run it from inside the dan node.

**Comment by javieramirez1** [comment]
```
[root@c83f2-dan9 noobaa-core]#  bin/node src/tools/s3cat --endpoint http://172.20.100.60:6001 --access_key Nk6lx6msTmKXUM04Chu9 --secret_key CzD+xVImh3WocHG1Ag670S+WG0+lqZ3GZMrQ1BMD --bucket bucket1 --upload --size $((1024*1024)) --part_size 1024 --concur 1
Uploading upload-lxjbbjpb from generated data of size 1.0 TB
UPLOAD ERROR: UnknownEndpoint: Inaccessible host: `172.20.100.60' at port `6001'. This service may not be available in the `us-east-1' region.
    at Request.ENOTFOUND_ERROR (/usr/local/noobaa-core/node_modules/aws-sdk/lib/event_listeners.js:611:46)
    at Request.callListeners (/usr/local/noobaa-core/node_modules/aws-sdk/lib/sequential_executor.js:106:20)
    at Request.emit (/usr/local/noobaa-core/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
    at Request.emit (/usr/local/noobaa-core/node_modules/aws-sdk/lib/request.js:686:14)
    at error (/usr/local/noobaa-core/node_modules/aws-sdk/lib/event_listeners.js:443:22)
    at ClientRequest.<anonymous> (/usr/local/noobaa-core/node_modules/aws-sdk/lib/http/node.js:100:9)
    at ClientRequest.emit (node:events:518:28)
    at ClientRequest.emit (node:domain:488:12)
    at Socket.socketErrorListener (node:_http_client:495:9)
    at Socket.emit (node:events:518:28) {
  code: 'UnknownEndpoint',
  region: 'us-east-1',
  hostname: '172.20.100.60',
  retryable: true,
  originalError: Error: connect ECONNREFUSED 172.20.100.60:6001
      at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1595:16) {
    errno: -111,
    code: 'NetworkingError',
    syscall: 'connect',
    address: '172.20.100.60',
    port: 6001,
    region: 'us-east-1',
    hostname: '172.20.100.60',
    retryable: true,
    time: 2024-06-17T18:32:16.380Z
  },
  time: 2024-06-17T18:32:16.380Z
```

**Comment by guymguym** [comment]
try `--endpoint http://localhost:6001`

**Comment by guymguym** [comment]
and check that your config.json has `"ALLOW_HTTP": true,`

**Comment by guymguym** [comment]
@javieramirez1 I managed to upload 1TB object with s3cat tool - does it upload for you now?
```
> bin/node src/tools/s3cat ...
...
...
upload done. 215 MB/sec

> ls -lh /gpfs/fs1/bucket1/
-rw-rw----. 1 root root 1.0T Jun 17 15:12 upload-lxj9v2bv
```

**Comment by javieramirez1** [comment]
It's in progress, it's not finished yet.
```
Upload Speed: 0.0 MB/sec (average 330.0)
Upload Speed: 476.7 MB/sec (average 330.4)
Upload Speed: 0.0 MB/sec (average 330.0)
Upload Speed: 487.4 MB/sec (average 330.4)
Upload Speed: 0.0 MB/sec (average 330.0)
Upload Speed: 489.5 MB/sec (average 330.4)
Upload Speed: 0.0 MB/sec (average 330.0)
Upload Speed: 543.5 MB/sec (average 330.4)
Upload Speed: 0.0 MB/sec (average 330.1)
Upload Speed: 517.4 MB/sec (average 330.5)
Upload Speed: 0.0 MB/sec (average 330.1)
Upload Speed: 487.9 MB/sec (average 330.5)
Upload Speed: 0.0 MB/sec (average 330.1)
Upload Speed: 483.2 MB/sec (average 330.5)
Upload Speed: 0.0 MB/sec (average 330.1)
Upload Speed: 494.7 MB/sec (average 330.5)
Upload Speed: 0.0 MB/sec (average 330.1)
Upload Speed: 488.1 MB/sec (average 330.5)
Upload Speed: 0.0 MB/sec (average 330.1)

```

**Comment by guymguym** [comment]
(BTW, unrelated to this large size issue, when I use concurrency `--upload --size $((100*1024)) --part_size 32 --concur 16` I managed to reproduce the "cancelled due to ctime change" issue - https://github.com/noobaa/noobaa-core/issues/7833, so for now please keep testing with large part size and concur of 1, and I will reopen the other issue, fyi @romayalon)

**Comment by javieramirez1** [comment]
[root@c83f2-dan9 bucket1]#  ls -lh /gpfs/remote_rw_fs0/buckets_1/bucket1/.noobaa-nsfs_*/multipart-uploads/*/parts-size-*
-rw-rw----. 1 bin bin 830G Jun 17 15:33 /gpfs/remote_rw_fs0/buckets_1/bucket1/.noobaa-nsfs_666730c4482e5fed3d54547d/multipart-uploads/87e8575e-b6c0-4611-bcec-d88aff50d1f2/parts-size-1073741824

**Comment by javieramirez1** [comment]
Upload Speed: 1023.0 MB/sec (average 330.7)
Upload Speed: 0.0 MB/sec (average 330.6)
upload done. 331 MB/sec

**Comment by javieramirez1** [comment]
Ok then I would use s3cat now from my app nodes, or what would I do?

**Comment by guymguym** [comment]
Hey @javieramirez1 

TLDR -- yes, s3cat is a convenient test tool and we can extend it to do more as needed, it uses the standard aws sdk for js, and you can install it with the [noobaa-core 5.15.4 rpm](https://noobaa-core-rpms.s3.amazonaws.com/noobaa-core-5.15.4-20240616-5.15.el8.x86_64.rpm) on the app nodes too. 

Few points:
1. As you increase the object size you also have to increase the part_size because the maximum number of parts is not allowed to exceed 10,000, which means that part_size of 1 GB can only get you up to object size of 10000 GB = 9.76 TB (which is a bit under 10 TB).
2. We can open an issue to [aws-cli](https://github.com/aws/aws-cli) about this problem. It's probably coming from its deps [botocore](https://github.com/boto/botocore) or [s3transfer](https://github.com/boto/s3transfer). It might get rejected if they do not plan to support it but still worth reporting imho.
3. We should verify extra-large uploads also with other s3 clients that can do above 1TB because it's not healthy to verify a server only with your own test client.
4. We can try aws s3api commands in a script which should probably work and be easy to use for testing - see [create-mpu](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/s3api/create-multipart-upload.html), [upload-part](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/s3api/upload-part.html), [complete-mpu](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/s3api/complete-multipart-upload.html).
5. Two other popular cli clients to try are [rclone](https://rclone.org/) and [mc](https://github.com/minio/mc).
6. I wonder which real world S3 apps use extra-large objects - for example is presto parquet using [hive connector](https://prestodb.io/docs/current/connector/hive.html#amazon-s3-configuration) can use such object sizes, see [this question](https://stackoverflow.com/questions/55034284/optimal-maximum-parquet-file-size-in-s3).

**Comment by guymguym** [comment]
Closing this issue for now, since it seems to be a client side issue only. Thanks @javieramirez1 !
